### PR TITLE
perf(layers): skip _clear_losses recursion when no losses ever registered

### LINKED ISSUE #23292: _clear_losses walks every sublayer even when no losses exist
`Layer._clear_losses` recurses into every sublayer on every forward call, walking the full sublayer tree even when the model has never called `add_loss` and has no `activity_regularizer` anywhere in it. This runs unconditionally as part of the per-call hot path in `keras/src/layers/layer.py`.

The walk is redundant on every call where it hasn't happened yet: whether a model can ever produce losses is fixed once its graph is built (`add_loss` calls happen at layer construction/call-definition time, not per forward step), so re-checking the whole tree on every single call for something that has never occurred wastes work for the entire lifetime of most models, which never register a loss at all.

Fixed by #23298: skip the recursive clear until the first `add_loss` call anywhere flips a process-global flag; once that happens, normal per-call clearing resumes as before.

Part of the per-call Python-dispatch overhead series in #22561.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23292

## Why this was closed

This change showed no measurable macro benefit on either axis. Its isolated
eager effect is within measurement noise on the models tested, and it is
compile-neutral: 0 graph-break delta with deterministic dynamo decomposition
(see the per-PR compile scorecard). It was backed only by a deterministic
micro-benchmark (a 1 → 0 reduction in recursive-clear call count on models
that never register a loss), not by a measured macro speedup.

torch.compile graph breaks across the #22561 series (this PR does not change the count)

## Summary

`_get_call_context` unconditionally calls `self._clear_losses()` on every
outermost layer/model call, recursing into every sublayer's `_losses` /
`_loss_ids` even when the model never calls `add_loss`. This adds a
process-global flag, `_LOSSES_REGISTERED`, set by `add_loss` (the sole
writer of `_losses`/`_loss_ids`, including the `activity_regularizer`
path), and skips the recursive clear until it's set. On models that never
register a loss, this drops the walk from 1 call/forward-pass to 0.

## Why it's safe

The flag is one-way (`False -> True`, never reset) and flipped
synchronously at the top of `add_loss`, so there's no window where a loss
exists but the flag still reads `False` — skipping the clear while unset
cannot skip clearing anything real. Once any loss is registered anywhere
in the process, behavior is byte-identical to the old always-clear path.
Change is confined to `keras/src/layers/layer.py` above backend dispatch,
no `backend.backend()` guards, so it applies uniformly across backends.

PR #23298: master vs PR alone vs PR+parents, Keras torch GPU eager, ms

Isolated GPU wall-clock lands as noise (expected — one skipped Python-level
recursion on toy models where GPU kernel launch dominates); the
deterministic proof is the 1 → 0 call-count reduction. The "+ all 13 series
PRs" column shows composability with the rest of #22561 (1.25–1.38×) and
isn't attributable to this PR alone — it has no parent PRs and no
sequencing dependency with the rest of the series.

## Tests

`layer_test.py`: flag stays unset with no losses (no `_clear_losses` call),
flips permanently on first `add_loss` (direct and via
`activity_regularizer`), and loss collection/clearing across nested
sublayers still works once set. Full `layers/` suite clean on torch
(matches baseline pass/skip counts); `layer_test.py` clean on tensorflow
and jax.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/layers/layer.py [hunk 1]
```diff
@@ -67,6 +67,16 @@
...
 
 
+# Process-global, set-once flag. Stays `False` until the first `add_loss`
+# call anywhere in the process. While it is `False`, no layer's `_losses`
+# list or `_loss_ids` set can be non-empty (the only writer is `add_loss`,
+# which flips this flag before doing so), so the recursive
+# `_clear_losses()` walk in `_get_call_context` is provably a no-op and can
+# be skipped entirely. Once set, it is never unset, and behavior reverts to
+# exactly the prior (always-clear) semantics for the rest of the process.
+_LOSSES_REGISTERED = False
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/layer.py:77]: The flag is intentionally process-global and monotonic, because the state it guards (`layer._losses` / `_loss_ids`) is itself process-global: plain object attributes, shared across threads and untouched by `clear_session()`. The skip is only sound while "flag is False" implies "no losses exist anywhere in the process", so the flag must be at least as shared and long-lived as the losses themselves. Moving it into `global_state` (thread-local, reset by `clear_session()`) breaks that invariant and would under-clear relative to master in two cases:

1. `clear_session()` replaces the thread-local tracker but does not touch existing layers' `_losses` (layers remain usable after `clear_session()` -- its own docstring example relies on that). A layer that registered a loss before `clear_session()` would skip the clear on its next call and accumulate: `layer.losses` goes from 1 entry to 2. On master it stays at 1.
2. A model that registered eager losses on one thread and is then called from a fresh worker thread (where the thread-local flag defaults to False) would skip clearing the shared `_losses` list and append to it. On master the worker's call clears first.

On thread-safety of the current form: it is a set-once monotonic bool, and `add_loss` flips it before appending the first loss, so under the GIL no reader can observe False after any loss exists. The only possible race outcome is reading True early, which just performs a redundant (safe) clear. Persisting across `clear_session()` is likewise deliberate: staying True only forfeits the optimization and reverts to the exact pre-change always-clear behavior, so it can never produce incorrect results, whereas a resettable flag can.

File: keras/src/layers/layer.py [hunk 2]
```diff
@@ -1275,6 +1285,8 @@ def call(self, x):
                     "`add_loss()` can only be called from inside `build()` or "
                     f"`call()`, on a tensor input. Received invalid value: {x}"
                 )
+        global _LOSSES_REGISTERED
+        _LOSSES_REGISTERED = True
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/layer.py:1289]: Keeping the module global here -- see the discussion on the flag declaration above. The flag must survive `clear_session()` and be visible across threads to keep the skip provably a no-op, and a `global_state` attribute provides neither. Note the write is deliberately placed before the losses are appended, so a concurrent reader can never see the flag unset while a loss exists.

File: keras/src/layers/layer.py [hunk 3]
```diff
@@ -1690,7 +1702,8 @@ def _get_call_context(self):
             global_state.set_global_attribute(
                 "current_call_ctx", layer_call_ctx
             )
-            self._clear_losses()
+            if _LOSSES_REGISTERED:
+                self._clear_losses()
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/layer.py:1706]: Same rationale as on the declaration thread: gating on a thread-local, session-reset flag would skip clearing `_losses` lists that were populated on another thread or before a `clear_session()` call -- both of which master clears today. Since `_losses` lives on the layer objects (process-shared, surviving `clear_session()`), the guard must be process-global and monotonic; that way it can only ever over-clear, never under-clear.

File: keras/src/layers/layer_test.py [hunk 4]
```diff
@@ -637,6 +638,148 @@ def call(self, x):
...
+    def test_clear_losses_walk_skipped_when_no_losses_ever_registered(self):
+        # `_get_call_context` should not walk the sublayer tree calling
+        # `_clear_losses()` on a model that has never registered a loss
+        # anywhere in the process. Once any layer anywhere calls
+        # `add_loss`, the module-level flag flips permanently and the walk
+        # always runs again (conservative over-clearing, never
+        # under-clearing).
+        original_flag = layer_module._LOSSES_REGISTERED
+        try:
+            layer_module._LOSSES_REGISTERED = False
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/layer_test.py:650]: The tests match the implementation, which is staying with the process-global module-level flag -- see the rationale on the declaration thread above. `layer_module._LOSSES_REGISTERED` is the actual mechanism being exercised; switching the tests to read/write `global_state` would test a different (and, per that discussion, incorrect) design rather than the one in this PR.

**Comment by pctablet505** [comment]
Closing this one. The isolated effect is within measurement noise on the models tested — it's a deterministic micro-optimization (skips the no-op _clear_losses recursion when no losses are ever registered) without a measurable end-to-end speedup, and it's compile-neutral. Trimming the #22561 series to the changes with macro impact to keep review focused.
